src/cockpit/peer.py

```diff
@@ -142,3 +142,41 @@ def do_channel_control(self, channel: str, command: str, message: Dict[str, obje
 
     def do_channel_data(self, channel: str, data: bytes) -> None:
         self.write_channel_data(channel, data)
+
+
+class PeerRoutingRule(RoutingRule, PeerStateListener):
+    predicate: Callable[[Dict[str, object]], bool]
+    args: List[str]
+    env: Dict[str, str]
+    peer: Optional[Peer]
+
+    def __init__(self, router: Router, args: List[str], env: Dict[str, str], predicate: Callable[[Dict[str, object]], bool]):
+        super().__init__(router)
+
+        self.args = args
+        self.env = env
+        self.predicate = predicate
+        self.peer = None
+
+    def apply_rule(self, options: Dict[str, object]) -> Optional[Peer]:
+        if not self.predicate(options):
+            return None
+
+        if self.peer is None:
```

Comment:
Is this supposed to be idempotent?

Reply:
Looking at src/cockpit/router.py, this is called for every channel open, so this acts like a cache and it should be idempotent.